enhancements/etcd/storage-migration-for-etcd-encryption.md

```diff
@@ -0,0 +1,182 @@
+
+## Implementation Details/Notes/Constraints
+
+* Create a `openshift-kube-storage-version-migrator-operator`, managed by the CVO, to manage the resources needed to run the `kube-storage-version-migrator`.
+* Initiate a storage migration of resources, even if the storage version has not changed.
+* Monitor the progress of storage migration of resources.
+* Enable the `kube-apiserver-operator` to utilize the `StorageVersionMigration` API when managing etcd encryption keys.
+* Enable the `openshift-apiserver-operator` to utilize the `StorageVersionMigration` API when managing etcd encryption keys.
+* Inform the upstream `kube-storage-version-migrator` of our usage and our success/failure to gain influence for future decisions and create tests to ensure our use-cases.
+
+New control loop added to kube-apiserver-operator & openshift-apiserver-operator:
+
+```
+for { // control loop
+
+    key := getNextWriteKey()
+
+    for resource := range encryptedResources {
+        migration := getStorageVersionMigrationFor(resource, key)
+        if migration not found {
+            createStorageVersionMigrationFor(resource, key)
+        }
+    }
+
+    if migrationsAreCompleteFor(key) {
+        setOperatorStatusPending(false)
+        enableWriteKey(key)
+    } else {
+        setOperatorStatusPending(false)
+    }
+        
+}
+```
+
+## Risks and Mitigations
+
+* kube-storage-version-migrator is currently in alpha with a clear plan to beta.
+
+* If encryption keys change while api-servers are a different levels, forcing a storage migration might render some api-servers unable to read the etcd data. To mitigate, we don’t request storage migrations unless all control planes are at the same level.
+
+* In order to insure that storage migrations initiated by other components. For etcd encryption, for example, we distribute and maintain multiple read keys allowing for a storage migration to “occur” even when not all control planes have the latest shared write key.
+
+## Design Details
+
+#### How do operators communicate the need for migration?
+
+They create a `StorageVersionMigration` resource for the GroupVersionResource they wish to migrate.
+
+#### How do operators get progress info?
+
+ By examining `StorageVersionMigration.status.condition[@type='Succeeded|Running']`.
+
+#### What happens if multiple operators need migration?
+
+Each operator creates thier own request (`StorageVersionMigration`) for a specific GVR to be migrated. Each operator examines thier corresponding request for progress info. If multiple migration requests are made for the same GVR, those will be exected serially.
+
+#### Is migration cpu or memory bound on the migrator side or just heavy for etcd?
+
+>  TODO
+
+#### Do we need throttling?
+
+If we do, it would be a function of the upstream controller. Some throttling is currently done by the upstream controller.
+
+#### Do we want or need CRD migration as well which might mean to have a user-facing API.
+
+> TODO
```

Comment:
The API is user facing. So I assume it includes CRD support.

Of course, it is alpha and therefore unsupported.